### PR TITLE
S3 Cleanup in SMM CPU Driver

### PR DESCRIPTION
**Pull request by jiaxinwu**
# S3 Cleanup in SMM CPU Driver

This PR belongs to the code cleanup & refinement.

There are 3 commits in the PR:
1. clean unused PCD for S3
This patch is to clean the PcdCpuFeaturesInitOnS3Resume since it's unused after commit 077760fe

2.  Clean redundant SmmS3Cr3 Init
The SmmS3Cr3 is only used by S3Resume PEIM to switch CPU from 32bit to 64bit, it should be the CR3 for Non-SMM environment and init by InitSmmS3Cr3 function. No need set to SMM CR3.

3. Avoid use global variable in InitSmmS3Cr3
This patch is to avoid use global variable in InitSmmS3Cr3. No function impact.

No function impact for all patches in the PR.

## How This Was Tested
Pass the build & boot & CI check.

### CODE REVIEW

File: UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c [hunk 1]
```diff
@@ -233,7 +233,6 @@ InitSmmS3ResumeState (
     }
 
     SmmS3ResumeState->SmmS3Cr0 = (UINT32)AsmReadCr0 ();
-    SmmS3ResumeState->SmmS3Cr3 = Cr3;
```
Comment (lzeng14) [inline UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c:236]: could the input parameter of InitSmmS3ResumeState be also removed ?

Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c:236]: agree

File: UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c [hunk 2]
```diff
@@ -233,7 +231,6 @@ InitSmmS3ResumeState (
...
```
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c:236]: Jiaxin, can you remove mSmmS3ResumeState?

Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c:236]: synced with ray, keep the mSmmS3ResumeState since it's needed.

**Comment by leiflindholm** [comment]
This PR has no description. Please address. And please fix the title.

File: UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c [hunk 3]
```diff
@@ -243,8 +243,9 @@ InitSmmS3ResumeState (
 
     //
     // Patch SmmS3ResumeState->SmmS3Cr3
+    // The SmmS3Cr3 is only used by S3Resume PEIM to switch CPU from 32bit to 64bit
     //
-    InitSmmS3Cr3 ();
+    InitSmmS3Cr3 (&mSmmS3ResumeState->SmmS3Cr3);
```
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c:248]: please use SmmS3ResumeState, not mSmmS3ResumeState

File: UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/SmmProfileArch.c [hunk 4]
```diff
@@ -12,13 +12,17 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
   Create SMM page table for S3 path.
 
+  @param[out] Cr3    The base address of the page tables.
+
 **/
 VOID
 InitSmmS3Cr3 (
-  VOID
+  OUT UINT64  *Cr3
   )
```
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/SmmProfileArch.c:21]: CR3 is 32bit in 32bit env. Using UINT64 is strange.

Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/SmmProfileArch.c:21]: CR3 is 32bit in 32bit env. Using UINT64 is strange.

Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/SmmProfileArch.c:21]: ok, i will change to UINTN

File: UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c (hunk 3 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c:248]: agree

**Review by niruiyu — APPROVED** [review]

**Pull request merged**